### PR TITLE
Default Load Screen Support

### PR DESCRIPTION
**Pull request by vmednonogov**
Ticket: 15138 (comment Jul/24 from Lloyd)
## What does this PR do?

Added support for Default Load Screen

## How was this PR tested?

Verified on PC
Waiting Jenkins build for iOS

### CODE REVIEW

File: Code/Legacy/CryCommon/LoadScreenComponent.cpp [hunk 1]
```diff
@@ -42,10 +42,6 @@ namespace
 
 #if defined(CARBONATED)
     static const char* const s_levelLoadScreenUiCanvasPath = "level_load_screen_uicanvas_path";
-
-    static const char* const s_globalLevelFixedFpsCvarName = "global_level_load_screen_sequence_fixed_fps";
-    static const char* const s_globalLevelMaxFpsCvarName = "global_level_load_screen_max_fps";
-    static const char* const s_globalLevelMinimumLoadTimeCvarName = "global_level_load_screen_minimum_time";
```
Comment (Twolewis) [inline Code/Legacy/CryCommon/LoadScreenComponent.cpp]: Why was this removed?  We -should- be defining the global load screen (as a default)

File: Code/Legacy/CryCommon/LoadScreenComponent.cpp [hunk 2]
```diff
@@ -272,18 +268,18 @@ void LoadScreenComponent::LevelStart()
...
 
             if (levelPath != "")
             {
                 LoadConfigSettings(s_levelFixedFpsCvarName, s_levelMaxFpsCvarName, s_levelMinimumLoadTimeCvarName);
             }
             else
             {
-                LoadConfigSettings(s_globalLevelFixedFpsCvarName, s_globalLevelMaxFpsCvarName, s_globalLevelMinimumLoadTimeCvarName);
+                // this level does not contain the load screen
+                return;
```
Comment (Twolewis) [inline Code/Legacy/CryCommon/LoadScreenComponent.cpp]: If the level does not contain a loading screen, then we should load the default from s_globalLevel... 

Not sure why this was removed?

File: Code/Legacy/CrySystem/LevelSystem/SpawnableLevelSystem.cpp [hunk 3]
```diff
@@ -542,15 +539,53 @@ namespace LegacyLevelSystem
...
+        }
+
+        const AZStd::string levelNameStr(levelName);
+        for (const auto& nameForExclude : levelNamesForExclude)
+        {
+            if (levelNameStr.contains(nameForExclude))
+            {
+                return false; // level does not have a load screen
+            }
+        }
```
Comment (Twolewis) [inline Code/Legacy/CrySystem/LevelSystem/SpawnableLevelSystem.cpp]: We should not need to do this at all.  If you're already checking whether or not the file exists, we can make a simple rule that says:
For every level is that is loaded, the following process runs:
* Loads the default level config (to set default values)
* IF the level-specific config exists, load it (overriding the values from the default)

This will allow us to add additional per-level configuration settings.  Concern is that if Level A sets "some_var", that value will persist to Level B.  Maybe we can enforce something where only settings in the default level configuration be overridden.  So:

default_level_config.cfg
some_var_1= 1
some_var_2 =2

urban/urban/level.cfg
some_var_1= 10 // Overrides some_var_1 
some_var_2 = 20 // Overrides some_var_2
some_var_3 = 30  // Does nothing.  some_var_3 is not in default_level_config.cfg

File: Code/Legacy/CrySystem/SystemInit.cpp [hunk 4]
```diff
@@ -1263,6 +1263,10 @@ void CSystem::CreateSystemVars()
         0,
         "Minimum amount of time to show the level load screen. Important to prevent short loads from flashing the load screen. 0 means "
         "there is no limit.");
+#if defined(CARBONATED)
+    REGISTER_STRING("default_load_screen_config_file", "", 0, "Default Load Screen Config File.");
```
Comment (Twolewis) [inline Code/Legacy/CrySystem/SystemInit.cpp]: Per my comment above, rename to "default_level_config_file"

File: Code/Legacy/CrySystem/SystemInit.cpp [hunk 5]
```diff
@@ -1263,6 +1263,10 @@ void CSystem::CreateSystemVars()
...
+    REGISTER_STRING("list_of_levels_without_load_screen", "", 0, "Comma separated list of levels without the level load screen.");
```
Comment (Twolewis) [inline Code/Legacy/CrySystem/SystemInit.cpp]: Per my earlier comment, remove this, and determine whether or not a level has a default based on the presence of their levels/level/level.cfg file.

**Review by Twolewis — CHANGES_REQUESTED** [review]

File: Code/Legacy/CrySystem/LevelSystem/SpawnableLevelSystem.cpp (hunk 3 above)
Comment (vmednonogov) [reply Code/Legacy/CrySystem/LevelSystem/SpawnableLevelSystem.cpp]: I did absolutely the same as you described above.
But we have two special non-battle levels which are loaded at the same code place with other levels:
1) "launch"
2) "worldmap"

I can't hardcoded their names into engine code to disable loading screen for them. And I added "list_of_levels_without_load_screen" cvar.
Ok, I will try to do something to avoid using of "list_of_levels_without_load_screen" cvar.
In the past I tried to enable the loading screen for "worldmap" level and it worked but looked slightly unusual. 
For "launch" level the loading screen crashed on tick processing and progress bar was not calculated, due to many things were being initialized at the same time and they were not ready.

Comment (Twolewis) [reply Code/Legacy/CrySystem/LevelSystem/SpawnableLevelSystem.cpp]: I see.  In the LY game we had a loading screen going to the worldmap and launch.  Understand that there's issues with the launch screen, but worldmap should have one too.

instead of this list that's maintained somewhere, we can just use the level loading screen cvars - if the path is overriden to be empty, no loading screen.

so then the launch screen can define an empty path (no loading screen), everything else will use the default, and specific levels can override if they want to use their own.

this eliminates the need to maintain a list

File: Code/Legacy/CryCommon/LoadScreenComponent.cpp (hunk 1 above)
Comment (vmednonogov) [reply Code/Legacy/CryCommon/LoadScreenComponent.cpp]: They are described in code but never used in LY game.
Ok, I have returned them back.

File: Code/Legacy/CryCommon/LoadScreenComponent.cpp (hunk 2 above)
Comment (vmednonogov) [reply Code/Legacy/CryCommon/LoadScreenComponent.cpp]: s_globalLevelXXX cvars are described in code but never used in LY game.
Ok, I have returned them back.

File: Code/Legacy/CrySystem/SystemInit.cpp (hunk 4 above)
Comment (vmednonogov) [reply Code/Legacy/CrySystem/SystemInit.cpp]: Renamed

File: Code/Legacy/CrySystem/SystemInit.cpp (hunk 5 above)
Comment (vmednonogov) [reply Code/Legacy/CrySystem/SystemInit.cpp]: Removed

File: Code/Legacy/CrySystem/LevelSystem/SpawnableLevelSystem.cpp (hunk 3 above)
Comment (vmednonogov) [reply Code/Legacy/CrySystem/LevelSystem/SpawnableLevelSystem.cpp]: Removed the list with excluded levels.
Added a simple missed checking to prevent crash during game loading (Added gEnv->pGame for non-null verification). It works on PC. Waiting Jenkins build for iOS verification.

Comment (vmednonogov) [reply Code/Legacy/CrySystem/LevelSystem/SpawnableLevelSystem.cpp]: @Twolewis : iOS verified via Build #26528 from Jenkins. 
The game does not crash on the launch. Worldmap progress bar appears.

**Review by Twolewis — APPROVED** [review]
